File: src/molexp/api/routes/projects.py
"""Project routes for MolExp API."""

from __future__ import annotations

from fastapi import APIRouter, Depends

from ..dependencies import get_workspace
from ..exceptions import ProjectNotFoundError
from ..schemas import MessageResponse, ProjectCreateRequest, ProjectResponse

router = APIRouter(prefix="/api/projects", tags=["projects"])


@router.get("", response_model=list[ProjectResponse])
def list_projects(workspace=Depends(get_workspace)) -> list[ProjectResponse]:
    """List all projects."""
    projects = workspace.list_projects()
    return [ProjectResponse.from_model(p) for p in projects]


@router.get("/{project_id}", response_model=ProjectResponse)
def get_project(project_id: str, workspace=Depends(get_workspace)) -> ProjectResponse:
    """Get project details."""
    project = workspace.get_project(project_id)
    if not project:
        raise ProjectNotFoundError(project_id)

    experiments = workspace.list_experiments(project_id)
    return ProjectResponse.from_model(project, experiment_count=len(experiments))


@router.post("", response_model=ProjectResponse, status_code=201)
def create_project(
    project: ProjectCreateRequest,
    workspace=Depends(get_workspace),
) -> ProjectResponse:
    """Create a new project."""
    new_project = workspace.create_project(
        project_id=project.project_id,
        name=project.name,
        description=project.description,
        owner=project.owner,
        tags=project.tags,
    )
    return ProjectResponse.from_model(new_project)


@router.delete("/{project_id}", response_model=MessageResponse)
def delete_project(
    project_id: str,
    workspace=Depends(get_workspace),
) -> MessageResponse:
    """Delete a project."""
    workspace.delete_project(project_id)
    return MessageResponse(message="Project deleted")

File: ui/mocks/handlers/projects.ts
/**
 * Mock handlers for Project API endpoints
 */

import { http, HttpResponse } from "msw";
import { getAllProjects, getProject, setProject, deleteProject, getAssetsByProject } from "../db";
import type { ApiProjectResponse } from "../../src/app/types";
import type { ProjectCreateRequest } from "../../src/api/generated/models/ProjectCreateRequest";

const API_BASE = "/api";

export const projectHandlers = [
    // GET /api/projects - List all projects
    http.get(`${API_BASE}/projects`, () => {
        const projects = getAllProjects();
        return HttpResponse.json(projects);
    }),

    // GET /api/projects - List all projects
    http.get(`${API_BASE}/projects/:id`, ({ params }) => {
        const { id } = params;
        const project = getProject(id as string);

        if (!project) {
            return HttpResponse.json(
                { message: `Project ${id} not found` },
                { status: 404 }
            );
        }

        return HttpResponse.json(project);
    }),

    // GET /api/projects/:id/assets - Get project assets
    http.get(`${API_BASE}/projects/:id/assets`, ({ params }) => {
        const assets = getAssetsByProject(params.id as string);
        return HttpResponse.json(assets);
    }),

    // POST /api/projects - Create new project
    http.post(`${API_BASE}/projects`, async ({ request }) => {
        const body = (await request.json()) as ProjectCreateRequest;

        const newProject: ApiProjectResponse = {
            id: body.name?.toLowerCase().replace(/\s+/g, "-") || `project-${Date.now()}`,
            projectId: body.name?.toLowerCase().replace(/\s+/g, "-") || `project-${Date.now()}`,
            name: body.name || "New Project",
            description: body.description || "",
            owner: body.owner || "molexp",
            tags: body.tags || [],
            config: {},
            created: new Date().toISOString(),
            experimentCount: 0,
        };

        setProject(newProject);

        return HttpResponse.json(newProject, { status: 201 });
    }),

    // DELETE /api/projects/:id - Delete project
    http.delete(`${API_BASE}/projects/:id`, ({ params }) => {
        const { id } = params;
        const deleted = deleteProject(id as string);

        if (!deleted) {
            return HttpResponse.json(
                { message: `Project ${id} not found` },
                { status: 404 }
            );
        }

        return HttpResponse.json({ message: "Project deleted" });
    }),
];
